query.QueryOperation Failed to serialize NormalResponse

```
Feb 10, 2017 6:56:29 AM com.hazelcast.map.impl.query.QueryOperation
SEVERE: [10.212.1.117]:5701 [HZ] [3.8-SNAPSHOT] Failed to serialize 'com.hazelcast.spi.impl.operationservice.impl.responses.NormalResponse'
com.hazelcast.nio.serialization.HazelcastSerializationException: Failed to serialize 'com.hazelcast.spi.impl.operationservice.impl.responses.NormalResponse'
    at com.hazelcast.internal.serialization.impl.SerializationUtil.handleSerializeException(SerializationUtil.java:73)
    at com.hazelcast.internal.serialization.impl.AbstractSerializationService.toBytes(AbstractSerializationService.java:154)
    at com.hazelcast.internal.serialization.impl.AbstractSerializationService.toBytes(AbstractSerializationService.java:125)
    at com.hazelcast.spi.impl.operationservice.impl.OutboundResponseHandler.send(OutboundResponseHandler.java:88)
    at com.hazelcast.spi.impl.operationservice.impl.OutboundResponseHandler.sendResponse(OutboundResponseHandler.java:64)
    at com.hazelcast.spi.Operation.sendResponse(Operation.java:338)
    at com.hazelcast.spi.impl.operationservice.impl.OperationRunnerImpl.sendResponse(OperationRunnerImpl.java:279)
    at com.hazelcast.spi.impl.operationservice.impl.OperationRunnerImpl.handleResponse(OperationRunnerImpl.java:270)
    at com.hazelcast.spi.impl.operationservice.impl.OperationRunnerImpl.run(OperationRunnerImpl.java:187)
    at com.hazelcast.spi.impl.operationservice.impl.OperationRunnerImpl.run(OperationRunnerImpl.java:401)
    at com.hazelcast.spi.impl.operationexecutor.impl.OperationThread.process(OperationThread.java:117)
    at com.hazelcast.spi.impl.operationexecutor.impl.OperationThread.run(OperationThread.java:102)
Caused by: com.hazelcast.nio.serialization.HazelcastSerializationException: Failed to serialize 'com.hazelcast.map.impl.query.AggregationResult'
    at com.hazelcast.internal.serialization.impl.SerializationUtil.handleSerializeException(SerializationUtil.java:73)
    at com.hazelcast.internal.serialization.impl.AbstractSerializationService.writeObject(AbstractSerializationService.java:251)
    at com.hazelcast.internal.serialization.impl.ByteArrayObjectDataOutput.writeObject(ByteArrayObjectDataOutput.java:370)
    at com.hazelcast.spi.impl.operationservice.impl.responses.NormalResponse.writeData(NormalResponse.java:91)
    at com.hazelcast.internal.serialization.impl.EnterpriseDataSerializableSerializer.write(EnterpriseDataSerializableSerializer.java:215)
    at com.hazelcast.internal.serialization.impl.EnterpriseDataSerializableSerializer.write(EnterpriseDataSerializableSerializer.java:56)
    at com.hazelcast.internal.serialization.impl.StreamSerializerAdapter.write(StreamSerializerAdapter.java:43)
    at com.hazelcast.internal.serialization.impl.AbstractSerializationService.toBytes(AbstractSerializationService.java:151)
    ... 10 more
Caused by: java.lang.NullPointerException
```

```
Feb 10, 2017 6:56:30 AM com.hazelcast.map.impl.query.MapQueryEngineImpl
WARNING: [10.212.1.117]:5701 [HZ] [3.8-SNAPSHOT] Could not get results
java.util.concurrent.ExecutionException: com.hazelcast.nio.serialization.HazelcastSerializationException: Failed to serialize 'com.hazelcast.spi.impl.operationservice.impl.responses
.NormalResponse'
	at com.hazelcast.spi.impl.operationservice.impl.InvocationFuture.resolveAndThrowIfException(InvocationFuture.java:88)
	at com.hazelcast.spi.impl.AbstractInvocationFuture.get(AbstractInvocationFuture.java:155)
	at com.hazelcast.map.impl.query.MapQueryEngineImpl.addResultsOfPredicate(MapQueryEngineImpl.java:177)
	at com.hazelcast.map.impl.query.MapQueryEngineImpl.dispatchQueryOnQueryThreads(MapQueryEngineImpl.java:153)
	at com.hazelcast.map.impl.query.MapQueryEngineImpl.doRunQueryOnQueryThreads(MapQueryEngineImpl.java:146)
	at com.hazelcast.map.impl.query.MapQueryEngineImpl.runQueryOnAllPartitions(MapQueryEngineImpl.java:125)
	at com.hazelcast.map.impl.query.MapQueryEngineImpl.execute(MapQueryEngineImpl.java:86)
	at com.hazelcast.map.impl.proxy.MapProxyImpl.aggregate(MapProxyImpl.java:783)
	at hzcmd.map.aggregate.Count.timeStep(Count.java:15)
	at remote.bench.marker.MetricsMarker.flatOut(MetricsMarker.java:52)
	at remote.bench.marker.MetricsMarker.bench(MetricsMarker.java:39)
	at remote.bench.BenchThread.call(BenchThread.java:38)
	at remote.bench.BenchThread.call(BenchThread.java:12)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
	at java.lang.Thread.run(Thread.java:745)
Caused by: com.hazelcast.nio.serialization.HazelcastSerializationException: Failed to serialize 'com.hazelcast.spi.impl.operationservice.impl.responses.NormalResponse'
	at com.hazelcast.internal.serialization.impl.SerializationUtil.handleSerializeException(SerializationUtil.java:73)
	at com.hazelcast.internal.serialization.impl.AbstractSerializationService.toBytes(AbstractSerializationService.java:154)
	at com.hazelcast.internal.serialization.impl.AbstractSerializationService.toBytes(AbstractSerializationService.java:125)
	at com.hazelcast.spi.impl.operationservice.impl.OutboundResponseHandler.send(OutboundResponseHandler.java:88)
	at com.hazelcast.spi.impl.operationservice.impl.OutboundResponseHandler.sendResponse(OutboundResponseHandler.java:64)
	at com.hazelcast.spi.Operation.sendResponse(Operation.java:338)
	at com.hazelcast.spi.impl.operationservice.impl.OperationRunnerImpl.sendResponse(OperationRunnerImpl.java:279)
	at com.hazelcast.spi.impl.operationservice.impl.OperationRunnerImpl.handleResponse(OperationRunnerImpl.java:270)
	at com.hazelcast.spi.impl.operationservice.impl.OperationRunnerImpl.run(OperationRunnerImpl.java:187)
	at com.hazelcast.spi.impl.operationservice.impl.OperationRunnerImpl.run(OperationRunnerImpl.java:401)
	at com.hazelcast.spi.impl.operationexecutor.impl.OperationThread.process(OperationThread.java:117)
	at com.hazelcast.spi.impl.operationexecutor.impl.OperationThread.run(OperationThread.java:102)
	at ------ submitted from ------.(Unknown Source)
	at com.hazelcast.spi.impl.operationservice.impl.InvocationFuture.resolve(InvocationFuture.java:114)
	at com.hazelcast.spi.impl.operationservice.impl.InvocationFuture.resolveAndThrowIfException(InvocationFuture.java:75)
	... 18 more
Caused by: com.hazelcast.nio.serialization.HazelcastSerializationException: Failed to serialize 'com.hazelcast.map.impl.query.AggregationResult'
	at com.hazelcast.internal.serialization.impl.SerializationUtil.handleSerializeException(SerializationUtil.java:73)
	at com.hazelcast.internal.serialization.impl.AbstractSerializationService.writeObject(AbstractSerializationService.java:251)
	at com.hazelcast.internal.serialization.impl.ByteArrayObjectDataOutput.writeObject(ByteArrayObjectDataOutput.java:370)
	at com.hazelcast.spi.impl.operationservice.impl.responses.NormalResponse.writeData(NormalResponse.java:91)
	at com.hazelcast.internal.serialization.impl.EnterpriseDataSerializableSerializer.write(EnterpriseDataSerializableSerializer.java:215)
	at com.hazelcast.internal.serialization.impl.EnterpriseDataSerializableSerializer.write(EnterpriseDataSerializableSerializer.java:56)
	at com.hazelcast.internal.serialization.impl.StreamSerializerAdapter.write(StreamSerializerAdapter.java:43)
	at com.hazelcast.internal.serialization.impl.AbstractSerializationService.toBytes(AbstractSerializationService.java:151)
	at com.hazelcast.internal.serialization.impl.AbstractSerializationService.toBytes(AbstractSerializationService.java:125)
	at com.hazelcast.spi.impl.operationservice.impl.OutboundResponseHandler.send(OutboundResponseHandler.java:88)
	at com.hazelcast.spi.impl.operationservice.impl.OutboundResponseHandler.sendResponse(OutboundResponseHandler.java:64)
	at com.hazelcast.spi.Operation.sendResponse(Operation.java:338)
	at com.hazelcast.spi.impl.operationservice.impl.OperationRunnerImpl.sendResponse(OperationRunnerImpl.java:279)
	at com.hazelcast.spi.impl.operationservice.impl.OperationRunnerImpl.handleResponse(OperationRunnerImpl.java:270)
	at com.hazelcast.spi.impl.operationservice.impl.Ope
```

https://github.com/hazelcast/hzCmd-bench/blob/zeta/src/main/java/hz/map/aggregate/Count.java

--- a/hazelcast/src/main/java/com/hazelcast/aggregation/impl/MaxAggregator.java
+++ b/hazelcast/src/main/java/com/hazelcast/aggregation/impl/MaxAggregator.java
@@ -45,6 +45,9 @@ public final class MaxAggregator<I, R extends Comparable> extends AbstractAggreg
     }
 
     private boolean isCurrentlyLessThan(R extractedValue) {
+        if (extractedValue == null) {
+            return false;
+        }
         return max == null || max.compareTo(extractedValue) < 0;
     }
 

--- a/hazelcast/src/main/java/com/hazelcast/aggregation/impl/MinAggregator.java
+++ b/hazelcast/src/main/java/com/hazelcast/aggregation/impl/MinAggregator.java
@@ -45,6 +45,9 @@ public final class MinAggregator<I, R extends Comparable> extends AbstractAggreg
     }
 
     private boolean isCurrentlyGreaterThan(R otherValue) {
+        if (otherValue == null) {
+            return false;
+        }
         return min == null || min.compareTo(otherValue) > 0;
     }
 

--- a/hazelcast/src/main/java/com/hazelcast/map/impl/query/AggregationResult.java
+++ b/hazelcast/src/main/java/com/hazelcast/map/impl/query/AggregationResult.java
@@ -55,10 +55,14 @@ public class AggregationResult implements Result<AggregationResult>, IdentifiedD
 
     @Override
     public void combine(AggregationResult result) {
-        if (partitionIds == null) {
-            partitionIds = new ArrayList<Integer>(result.getPartitionIds().size());
+        Collection<Integer> otherPartitionIds = result.getPartitionIds();
+        if (otherPartitionIds == null) {
+            return;
         }
-        partitionIds.addAll(result.getPartitionIds());
+        if (partitionIds == null) {
+            partitionIds = new ArrayList<Integer>(otherPartitionIds.size());
+        }
+        partitionIds.addAll(otherPartitionIds);
         aggregator.combine((result.aggregator));
     }
 
@@ -86,20 +90,25 @@ public class AggregationResult implements Result<AggregationResult>, IdentifiedD
 
     @Override
     public void writeData(ObjectDataOutput out) throws IOException {
-        out.writeObject(aggregator);
-        out.writeInt(partitionIds.size());
-        for (Integer partitionId : partitionIds) {
-            out.writeInt(partitionId);
+        int partitionSize = (partitionIds == null) ? 0 : partitionIds.size();
+        out.writeInt(partitionSize);
+        if (partitionSize > 0) {
+            for (Integer partitionId : partitionIds) {
+                out.writeInt(partitionId);
+            }
         }
+        out.writeObject(aggregator);
     }
 
     @Override
     public void readData(ObjectDataInput in) throws IOException {
-        this.aggregator = in.readObject();
-        int partitionIdsSize = in.readInt();
-        this.partitionIds = new ArrayList<Integer>(partitionIdsSize);
-        for (int i = 0; i < partitionIdsSize; i++) {
-            this.partitionIds.add(in.readInt());
+        int partitionSize = in.readInt();
+        if (partitionSize > 0) {
+            this.partitionIds = new ArrayList<Integer>(partitionSize);
+            for (int i = 0; i < partitionSize; i++) {
+                this.partitionIds.add(in.readInt());
+            }
         }
+        this.aggregator = in.readObject();
     }
 }

--- a/hazelcast/src/main/java/com/hazelcast/map/impl/query/MapQueryEngineImpl.java
+++ b/hazelcast/src/main/java/com/hazelcast/map/impl/query/MapQueryEngineImpl.java
@@ -35,6 +35,7 @@ import java.util.List;
 import java.util.Set;
 import java.util.concurrent.ExecutionException;
 import java.util.concurrent.Future;
+import java.util.logging.Level;
 
 import static com.hazelcast.util.ExceptionUtil.rethrow;
 
@@ -155,7 +156,7 @@ public class MapQueryEngineImpl implements MapQueryEngine {
             if (t.getCause() instanceof QueryResultSizeExceededException) {
                 throw rethrow(t);
             }
-            logger.warning("Could not get results", t);
+            logger.log(Level.FINE, "Could not get results", t);
         }
     }
 

--- a/hazelcast/src/main/java/com/hazelcast/map/impl/query/QueryResult.java
+++ b/hazelcast/src/main/java/com/hazelcast/map/impl/query/QueryResult.java
@@ -124,10 +124,14 @@ public class QueryResult implements Result<QueryResult>, IdentifiedDataSerializa
 
     @Override
     public void combine(QueryResult result) {
-        if (partitionIds == null) {
-            partitionIds = new ArrayList<Integer>(result.getPartitionIds().size());
+        Collection<Integer> otherPartitionIds = result.getPartitionIds();
+        if (otherPartitionIds == null) {
+            return;
         }
-        partitionIds.addAll(result.getPartitionIds());
+        if (partitionIds == null) {
+            partitionIds = new ArrayList<Integer>(otherPartitionIds.size());
+        }
+        partitionIds.addAll(otherPartitionIds);
         rows.addAll(result.getRows());
     }
 

--- a/hazelcast/src/test/java/com/hazelcast/aggregation/AggregationMemberBounceTest.java
+++ b/hazelcast/src/test/java/com/hazelcast/aggregation/AggregationMemberBounceTest.java
@@ -1,0 +1,62 @@
+package com.hazelcast.aggregation;
+
+import com.hazelcast.config.Config;
+import com.hazelcast.core.HazelcastInstance;
+import com.hazelcast.core.IMap;
+import com.hazelcast.test.HazelcastSerialClassRunner;
+import com.hazelcast.test.HazelcastTestSupport;
+import com.hazelcast.test.annotation.SlowTest;
+import com.hazelcast.test.bounce.BounceMemberRule;
+import org.junit.Before;
+import org.junit.Rule;
+import org.junit.Test;
+import org.junit.experimental.categories.Category;
+import org.junit.runner.RunWith;
+
+@RunWith(HazelcastSerialClassRunner.class)
+@Category(SlowTest.class)
+public class AggregationMemberBounceTest extends HazelcastTestSupport {
+    private static final int MEMBER_COUNT = 4;
+    private static final int DRIVER_COUNT = 4;
+    private static final int TEST_DURATION_SECONDS = 40;
+
+    private String mapName;
+
+    @Rule
+    public BounceMemberRule bounceMemberRule = BounceMemberRule.with(getConfig())
+            .clusterSize(MEMBER_COUNT)
+            .driverCount(DRIVER_COUNT).build();
+
+    protected Config getConfig() {
+        return new Config();
+    }
+
+    @Before
+    public void setup() {
+        mapName = randomMapName();
+        final HazelcastInstance steadyMember = bounceMemberRule.getSteadyMember();
+        IMap<Integer, AggregatorsSpecTest.Person> map = steadyMember.getMap(mapName);
+        AggregatorsSpecTest.populateMapWithPersons(map);
+    }
+
+    @Test
+    public void aggregationReturnsCorrectResultWhenBouncing() {
+        Runnable[] runnables = new Runnable[DRIVER_COUNT];
+        for (int i = 0; i < DRIVER_COUNT; i++) {
+            HazelcastInstance driver = bounceMemberRule.getNextTestDriver();
+            final IMap<Integer, AggregatorsSpecTest.Person> map = driver.getMap(mapName);
+            runnables[i] = new Runnable() {
+                @Override
+                public void run() {
+                    AggregatorsSpecTest.assertMinAggregators(map);
+                    AggregatorsSpecTest.assertMaxAggregators(map);
+                    AggregatorsSpecTest.assertSumAggregators(map);
+                    AggregatorsSpecTest.assertAverageAggregators(map);
+                    AggregatorsSpecTest.assertCountAggregators(map);
+                    AggregatorsSpecTest.assertDistinctAggregators(map);
+                }
+            };
+        }
+        bounceMemberRule.testRepeatedly(runnables, TEST_DURATION_SECONDS);
+    }
+}

--- a/hazelcast/src/test/java/com/hazelcast/aggregation/AggregatorsSpecTest.java
+++ b/hazelcast/src/test/java/com/hazelcast/aggregation/AggregatorsSpecTest.java
@@ -59,8 +59,6 @@ public class AggregatorsSpecTest extends HazelcastTestSupport {
             throw new IllegalArgumentException("node count < 1");
         }
 
-        TestHazelcastInstanceFactory factory = createHazelcastInstanceFactory(nodeCount);
-
         MapConfig mapConfig = new MapConfig()
                 .setName("aggr")
                 .setInMemoryFormat(InMemoryFormat.OBJECT);
@@ -70,11 +68,12 @@ public class AggregatorsSpecTest extends HazelcastTestSupport {
                 .setProperty(AGGREGATION_ACCUMULATION_PARALLEL_EVALUATION.getName(), String.valueOf(parallelAccumulation))
                 .addMapConfig(mapConfig);
 
+        TestHazelcastInstanceFactory factory = createHazelcastInstanceFactory(nodeCount);
         HazelcastInstance instance = factory.newInstances(config)[0];
         return instance.getMap("aggr");
     }
 
-    private static void assertMinAggregators(IMap<Integer, Person> map) {
+    public static void assertMinAggregators(IMap<Integer, Person> map) {
         assertEquals(Double.valueOf(1), map.aggregate(Aggregators.<Map.Entry<Integer, Person>>doubleMin("doubleValue")));
         assertEquals(Long.valueOf(1), map.aggregate(Aggregators.<Map.Entry<Integer, Person>>longMin("longValue")));
         assertEquals(Integer.valueOf(1), map.aggregate(Aggregators.<Map.Entry<Integer, Person>>integerMin("intValue")));
@@ -90,7 +89,7 @@ public class AggregatorsSpecTest extends HazelcastTestSupport {
         assertEquals("1", map.aggregate(Aggregators.<Map.Entry<Integer, Person>, String>comparableMin("comparableValue")));
     }
 
-    private static void assertMaxAggregators(IMap<Integer, Person> map) {
+    public static void assertMaxAggregators(IMap<Integer, Person> map) {
         assertEquals(Double.valueOf(999), map.aggregate(Aggregators.<Map.Entry<Integer, Person>>doubleMax("doubleValue")));
         assertEquals(Long.valueOf(999), map.aggregate(Aggregators.<Map.Entry<Integer, Person>>longMax("longValue")));
         assertEquals(Integer.valueOf(999), map.aggregate(Aggregators.<Map.Entry<Integer, Person>>integerMax("intValue")));
@@ -106,7 +105,7 @@ public class AggregatorsSpecTest extends HazelcastTestSupport {
         assertEquals("999", map.aggregate(Aggregators.<Map.Entry<Integer, Person>, String>comparableMax("comparableValue")));
     }
 
-    private static void assertSumAggregators(IMap<Integer, Person> map) {
+    public static void assertSumAggregators(IMap<Integer, Person> map) {
         assertEquals(Double.valueOf(499500.0d), map.aggregate(Aggregators.<Map.Entry<Integer, Person>>doubleSum("doubleValue")));
         assertEquals(Long.valueOf(499500), map.aggregate(Aggregators.<Map.Entry<Integer, Person>>longSum("longValue")));
         assertEquals(Long.valueOf(499500), map.aggregate(Aggregators.<Map.Entry<Integer, Person>>integerSum("intValue")));
@@ -126,7 +125,7 @@ public class AggregatorsSpecTest extends HazelcastTestSupport {
         assertEquals(Double.valueOf(499500), map.aggregate(Aggregators.<Map.Entry<Integer, Person>>floatingPointSum("bigDecimalValue")));
     }
 
-    private static void assertAverageAggregators(IMap<Integer, Person> map) {
+    public static void assertAverageAggregators(IMap<Integer, Person> map) {
         assertEquals(Double.valueOf(500.0d), map.aggregate(Aggregators.<Map.Entry<Integer, Person>>doubleAvg("doubleValue")));
         assertEquals(Double.valueOf(500.0d), map.aggregate(Aggregators.<Map.Entry<Integer, Person>>longAvg("longValue")));
         assertEquals(Double.valueOf(500.0d), map.aggregate(Aggregators.<Map.Entry<Integer, Person>>integerAvg("intValue")));
@@ -140,7 +139,7 @@ public class AggregatorsSpecTest extends HazelcastTestSupport {
         assertEquals(Double.valueOf(500.0d), map.aggregate(Aggregators.<Map.Entry<Integer, Person>>numberAvg("bigIntegerValue")));
     }
 
-    private static void assertCountAggregators(IMap<Integer, Person> map) {
+    public static void assertCountAggregators(IMap<Integer, Person> map) {
         assertEquals(Long.valueOf(999), map.aggregate(Aggregators.<Map.Entry<Integer, Person>>count("doubleValue")));
         assertEquals(Long.valueOf(999), map.aggregate(Aggregators.<Map.Entry<Integer, Person>>count("longValue")));
         assertEquals(Long.valueOf(999), map.aggregate(Aggregators.<Map.Entry<Integer, Person>>count("intValue")));
@@ -149,7 +148,7 @@ public class AggregatorsSpecTest extends HazelcastTestSupport {
         assertEquals(Long.valueOf(999), map.aggregate(Aggregators.<Map.Entry<Integer, Person>>count("comparableValue")));
     }
 
-    private static void assertDistinctAggregators(IMap<Integer, Person> map) {
+    public static void assertDistinctAggregators(IMap<Integer, Person> map) {
         assertEquals(map.project(Projections.<Map.Entry<Integer, Person>, Double>singleAttribute("doubleValue")),
                 map.aggregate(Aggregators.<Map.Entry<Integer, Person>, Double>distinct("doubleValue")));
         assertEquals(map.project(Projections.<Map.Entry<Integer, Person>, Long>singleAttribute("longValue")),
@@ -164,7 +163,7 @@ public class AggregatorsSpecTest extends HazelcastTestSupport {
                 map.aggregate(Aggregators.<Map.Entry<Integer, Person>, Comparable>distinct("comparableValue")));
     }
 
-    private static void populateMapWithPersons(IMap<Integer, Person> map) {
+    public static void populateMapWithPersons(IMap<Integer, Person> map) {
         for (int i = 1; i < 1000; i++) {
             map.put(i, new Person(i));
         }

